### PR TITLE
[Distributed Strategy] add separate_mtp_headloss config check

### PR DESCRIPTION
**Pull request by Wennie396**
### PR Category
User Experience

### PR Types
Improvements

### Description
对开关separate_mtp_headloss增加config的check逻辑，**打开的情况下如果不符合要求自动关闭**

### 是否引起精度变化
否

### CODE REVIEW

File: src/paddlefleet/transformer/transformer_config.py [hunk 1]
```diff
@@ -2364,3 +2364,79 @@ def __post_init__(self):
...
+                    vpp_degree = 1
+
+                num_empty_layers = self.num_empty_layers_add_in_head + max(
+                    0, self.num_empty_layers_add_in_tail - 1
+                )
+                total_layers = (
+                    self.num_hidden_layers + mtp_num_layers + num_empty_layers
+                )
+                denom = pp_degree * vpp_degree
+                if total_layers % denom != 0 or total_layers // denom != 1:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py:2427]: P1

这里把 `total_layers // (pp_degree * vpp_degree)` 限定为 1，会误关正常的多层 PP/VPP 模型。比如现有 MTP+PP 测试的 `num_hidden_layers=11`、`num_nextn_predict_layers=3`、头/尾空层 `2/3`、`pp=4`、`vpp=2`，按本校验得到 `total_layers=18`、`denom=8`，因此 `separate_mtp_headloss` 会被强制改回 False；但该功能本身就是把 head/loss 插入流水线，单个 stage 合法承载多个层，并不要求总层数等于 stage 数。请移除 `quotient == 1` 的限制，只校验实际分片所需的可除性/布局契约，并补充一个开启该开关的 PP/VPP 回归测试。

File: src/paddlefleet/transformer/transformer_config.py [hunk 2]
```diff
@@ -2364,3 +2364,79 @@ def __post_init__(self):
...
+        # separate_mtp_headloss validation.
+        if self.separate_mtp_headloss:
+            import warnings as _warnings
+
+            # Resolve the effective number of MTP layers, following the same
+            # logic used elsewhere in __post_init__ (see the csa branch above).
+            mtp_num_layers = (
+                self.mtp_num_layers
+                if self.mtp_num_layers > 0
+                else self.num_nextn_predict_layers
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py]: P1

这里用 `mtp_num_layers` 作为 `mtp_enabled` 的依据，但实际调用链并不支持这个替代字段：`gpt_builder` 通过 `_get_effective_mtp_layers` 可以构建 MTP，随后 `get_gpt_spec` 在 `separate_mtp_headloss` 分支却明确断言 `config.num_nextn_predict_layers > 0`，`MainLanguageLoss`/`MTPLanguageLoss` 也直接按该字段切分 labels。于是配置只设置 `mtp_num_layers > 0`、`num_nextn_predict_layers=0` 时，本校验会保留开关，建模阶段仍触发断言，而不是按描述自动关闭。请与实际消费者统一有效 MTP 层数（或在校验中同时拒绝该组合），并覆盖两字段组合的测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [transformer_config.py:2427](https://github.com/PaddlePaddle/PaddleFleet/pull/1719#discussion_r3764283238) | P1 | ✅ |
| 2 | [transformer_config.py:2377](https://github.com/PaddlePaddle/PaddleFleet/pull/1719#discussion_r3764286290) | P1 | ✅ |
| 3 | [PR 标题](https://github.com/PaddlePaddle/PaddleFleet/pull/1719) | P2 | ✅ |
| 4 | [test_ai_separate_mtp_headloss.py:128](https://github.com/PaddlePaddle/PaddleFleet/pull/1719#discussion_r3766355953) | P1 | ✅ |
| 5 | [test_configs.yaml:32](https://github.com/PaddlePaddle/PaddleFleet/blob/72277e48eb658f274724d4d29e7066f6386ecf5a/tests/test_configs.yaml#L32) | P1 | ✅ |

PR 标题已补充 `[Distributed Strategy]` 前缀。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/transformer_config.py (hunk 1 above)
Comment (Wennie396) [reply src/paddlefleet/transformer/transformer_config.py:2427]: 这个功能由于底层ShareLayerDesc的机制问题，当denom>1的时候会有问题，所以这里会assert掉，这个是必须做出的check

Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_config.py:2427]: 感谢说明。我核对了当前底层实现：`PipelineLayer.get_stage_from_index()` 会按 pp/vpp 分段把多个 layer 映射到同一 stage，`_construct_shared_comm()` 也按实际包含 `SharedLayerDesc` 的 stage 建立通信组。因此，仅凭 `denom > 1` 或 `total_layers // denom != 1` 还不能证明所有这类布局都不支持；当前条件会把已有的 11 层、3 层 MTP、头/尾空层 2/3、PP=4、VPP=2 配置直接关闭。

如果该功能确实只支持某个更窄的 `ShareLayerDesc` 布局，请在校验和注释中表达精确契约，并补充一个实际开启 `separate_mtp_headloss` 的 PP/VPP 回归测试，覆盖共享层所在 stage。当前我会保留这条 P1，直到该限制有可验证的布局依据。新提交已统一使用 `num_nextn_predict_layers`，第二条问题已解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/transformer_config.py (hunk 2 above)
Comment (Wennie396) [reply src/paddlefleet/transformer/transformer_config.py]: Done

File: tests/single_card_tests/ai_edited_test/config_and_utils/test_ai_separate_mtp_headloss.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import sys
+import unittest
+import warnings
+
+sys.path.insert(
```
Comment (risemeup1111) [inline tests/single_card_tests/ai_edited_test/config_and_utils/test_ai_separate_mtp_headloss.py]: P1

该测试通过两处 `sys.path.insert` 改写导入环境，并保留脚本入口；仓库单卡测试已由 `pytest -s <file>` 运行，测试应依赖正常安装或可编辑安装的 `paddlefleet`，否则会掩盖打包/导入问题且行为依赖目录层级。请移除两处路径注入和 `__main__` 入口，并放到已有的 `TransformerConfig` 测试模块，或按 pytest 的正常导入方式执行。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: tests/single_card_tests/ai_edited_test/config_and_utils/test_ai_separate_mtp_headloss.py (hunk 3 above)
Comment (risemeup1111) [reply tests/single_card_tests/ai_edited_test/config_and_utils/test_ai_separate_mtp_headloss.py]: 最新提交已移除两处 `sys.path.insert`，但当前测试文件仍保留 `if __name__ == "__main__": unittest.main()`（第 128-129 行）。仓库 CI 通过 pytest 运行测试，不需要脚本入口；请删除该入口后再关闭这条 P1。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: tests/single_card_tests/ai_edited_test/config_and_utils/test_ai_separate_mtp_headloss.py (hunk 3 above)
Comment (Wennie396) [reply tests/single_card_tests/ai_edited_test/config_and_utils/test_ai_separate_mtp_headloss.py]: Done

Comment (risemeup1111) [reply tests/single_card_tests/ai_edited_test/config_and_utils/test_ai_separate_mtp_headloss.py]: 当前 head `de0b95f7` 的文件仍保留第 128-129 行 `if __name__ == "__main__": unittest.main()`，因此该 P1 还未解决。请在新提交中删除脚本入口；删除后我会重新核验并更新 Board 状态。

File: src/paddlefleet/transformer/transformer_config.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_config.py:2427]: 这次新增的 PP/VPP 回归测试确实构造了 quotient>1，但 `_build_config()` 在 `GPTConfig` 构造完成后于第 92 行手动执行 `config.separate_mtp_headloss = True`，绕过了待审的 `__post_init__` 校验。因此它不能证明用户正常配置会被允许；当前 `transformer_config.py` 仍会在 `total_layers // denom != 1` 时自动关闭，原 P1 仍未解决。请移除该手动覆盖，让构造路径本身验证正常开启，或修正 guard 契约。

File: tests/multi_card_tests/pipeline_parallel/test_separate_mtp_headloss_pp.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    return config
+
+
+class TestSeparateMtpHeadlossPP(unittest.TestCase):
+    def setUp(self):
+        strategy = fleet.DistributedStrategy()
+        strategy.hybrid_configs = {
+            "dp_degree": 1,
+            "mp_degree": 2,
+            "pp_degree": PP_DEGREE,
```
Comment (risemeup1111) [inline tests/multi_card_tests/pipeline_parallel/test_separate_mtp_headloss_pp.py:104]: P1

该测试在第 102 行设置 PP=4，同时第 101 行设置 MP=2、第 106 行设置 EP=2；但仓库 `ci/multi-card_test.sh` 会将本文件匹配到 `tests/test_configs.yaml` 的 `tests/multi_card_tests/pipeline_parallel/*.py` 回退项，实际只启动 2 张卡。现有同类 PP=4/MP=2 的 MTP 测试则显式配置 8 张卡。当前文件没有专用配置，因此 `initialize_fleet` 会在不匹配的 world size 下失败，回归测试无法运行。请在 `tests/test_configs.yaml` 为本文件增加与并行配置一致的 8 卡条目，并纳入 CI。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: tests/multi_card_tests/pipeline_parallel/test_separate_mtp_headloss_pp.py (hunk 4 above)
Comment (risemeup1111) [reply tests/multi_card_tests/pipeline_parallel/test_separate_mtp_headloss_pp.py:104]: 最新提交已在 `tests/test_configs.yaml` 为该测试补充 8 卡专用配置，PP=4/MP=2 的资源不匹配问题已解决；Board 第 5 行已更新为 ✅。

File: src/paddlefleet/transformer/transformer_config.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_config.py:2427]: 最新测试已移除手动覆盖：`GPTConfig.__post_init__` 在 PP=4/VPP=2、8 个 seg-weight layers 的布局中会正常保留 `separate_mtp_headloss=True`，并实际构造 PipelineLayer 分段与 SharedLayerDesc 通信。该布局 P1 已解决，我会同步更新 Board。

**Review by risemeup1111 — COMMENTED** [review]
标题已补充 `[Distributed Strategy]` 前缀，当前未发现未解决项。

**Review by risemeup1111 — APPROVED** [review]

**Review by From00 — APPROVED** [review]
LGTM

**Comment by risemeup1111** [comment]
❌ Cherry-pick failed: Conflicts detected when cherry-picking to `release/0.4`. Please resolve manually.
